metaflow/plugins/kubernetes/kubernetes_jobsets.py

```diff
@@ -0,0 +1,745 @@
+    subdomain,
+    control_task_id=None,
+    worker_task_id=None,
+    replicas=1,
+    port=None,
+    num_parallel=None,
+    namespace=None,
+) -> dict:
+    master_port = port
+
+    job_spec = copy.deepcopy(job_spec)
+    job_spec.parallelism = 1
+    job_spec.completions = 1
+    job_spec.template.spec.set_hostname_as_fqdn = True
+    job_spec.template.spec.subdomain = subdomain
+
+    for idx in range(len(job_spec.template.spec.containers[0].command)):
+        if control_task_id in job_spec.template.spec.containers[0].command[idx]:
+            job_spec.template.spec.containers[0].command[idx] = (
+                job_spec.template.spec.containers[0]
+                .command[idx]
+                .replace(control_task_id, worker_task_id)
+            )
+        # CHECK FOR THE ubf_context in the command.
+        # Replace the UBF context to the one appropriately matching control/worker.
+        # Since we are passing the `step_cli` one time from the top level to one
+        # KuberentesJobSet, we need to ensure that UBF context is replaced properly
+        # in all the worker jobs.
+        if UBF_CONTROL in job_spec.template.spec.containers[0].command[idx]:
+            # Since all command will have a UBF_CONTROL, we need to replace the UBF_CONTROL
+            # with the actual UBF Context and also ensure that we are setting the correct
+            # split-index for the worker jobs.
+            split_index_str = "--split-index `expr $WORKER_REPLICA_INDEX + 1`"  # This set in the environment variables below
+            job_spec.template.spec.containers[0].command[idx] = (
+                job_spec.template.spec.containers[0]
+                .command[idx]
+                .replace(UBF_CONTROL, UBF_TASK + " " + split_index_str)
+            )
+
+    job_spec.template.spec.containers[0].env = job_spec.template.spec.containers[
+        0
+    ].env + _jobset_specific_env_vars(
+        client, jobset_main_addr, master_port, num_parallel
+    )
+
+    # Based on https://github.com/kubernetes-sigs/jobset/blob/v0.5.0/api/jobset/v1alpha2/jobset_types.go#L178
+    return dict(
+        name=job_name,
+        template=client.api_client.ApiClient().sanitize_for_serialization(
+            client.V1JobTemplateSpec(
+                metadata=client.V1ObjectMeta(
+                    namespace=namespace,
+                    # We don't set any annotations here
+                    # since they have been either set in the JobSpec
+                    # or on the JobSet level
+                ),
+                spec=job_spec,
+            )
+        ),
+        replicas=replicas,
+    )
+
+
+def _make_domain_name(
+    jobset_name, main_job_name, main_job_index, main_pod_index, namespace
+):
+    return "%s-%s-%s-%s.%s.%s.svc.cluster.local" % (
+        jobset_name,
+        main_job_name,
+        main_job_index,
+        main_pod_index,
+        jobset_name,
+        namespace,
+    )
+
+
+class KubernetesJobSet(object):
```

Comment:
It creates one control and one worker definition. The workers will have replicas set to `num_parallel-1`. All workers and control will leverage the jobset created before them.